qiskit_experiments/framework/backend_timing.py

```diff
@@ -0,0 +1,349 @@
+        does not provide it rather than just giving a ``TypeError`` about
+        ``NoneType``. As this raises an exception when ``dt`` is not set, it
+        likely should not be used by external code using
+        :class:`.BackendTiming`.
+
+        Raises:
+            QiskitError: The backend does not include a dt value.
+        """
+        if self._backend_data_dt is not None:
+            return float(self._backend_data_dt)
+
+        raise QiskitError("Backend has no dt value.")
+
+    def circuit_delay(self, time: float) -> Union[int, float]:
+        """Delay duration close to ``time`` and consistent with timing constraints
+
+        This method produces the value to pass for the ``duration`` of a
+        ``Delay`` instruction of a ``QuantumCircuit`` schedule so that the
+        delay fills the time until the next valid pulse, assuming the ``Delay``
+        instruction begins on a sample that is also valid for a pulse to begin
+        on.
+
+        The pulse timing constraints of the backend are considered in order to
+        give a number of samples closest to ``time`` plus however many more
+        samples are needed to get to the next valid sample for the start of a
+        pulse in a subsequent instruction. The least common multiple of the
+        pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        If :meth:`.BackendTiming.delay_unit` is ``s``, ``time`` is
+        returned directly. Typically, this is the case for a simulator where
+        converting to sample number is not needed.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples if :meth:`.BackendTiming.delay_unit`
+            is ``dt``. Other return ``time``.
+        """
+        if self.delay_unit == "s":
+            return time
+
+        return self.schedule_delay(time)
+
+    def schedule_delay(self, time: float) -> int:
+        """Valid delay value in samples to use in a pulse schedule for  ``time``
+
+        The pulse timing constraints of the backend are considered in order to
+        give a number of samples closest to ``time`` plus however many more
+        samples are needed to get to the next valid sample for the start of a
+        pulse in a subsequent instruction. The least common multiple of the
+        pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples to pass to a ``Delay`` instruction in
+            Qiskit pulse schedule
+        """
+
+        granularity = lcm(self._pulse_alignment, self._acquire_alignment)
+
+        samples = int(round(time / self._dt / granularity) * granularity)
+
+        return samples
+
+    def pulse_samples(self, time: float) -> int:
```

Comment:
Usually we input pulse time in dt. Is it possible to prepare another interface or second argument for unit of `time` ?

Reply:
That seems reasonable. What is an example of when you would have a duration in samples? I was thinking this would be used when doing something like spectroscopy or sweeping a gate length. Without a conversion function like this, it might be more natural to start from samples, but are there times one still wants to use samples when it is easy to convert from seconds?

For the signature, we could have `time` and `samples` arguments and raise if both or none are passed. Or we could have a separate method that just takes samples. Since there are several time/duration quantities being handled, I think the most important job for this class is to make the functions as clear as possible (that was part of why I separate circuit_delay and schedule_delay).

Reply:
I added `round_pulse_samples` in 3c11b2c as a separate method to keep the usage clear (maybe). Maybe having `pulse_samples` and `round_pulse_samples` is confusing. Trying to rename `pulse_samples` seems difficult though. `time_to_pulse_samples` is getting long. It might be better to have just `pulse_samples` and make it require a keyword argument, so you either have to do `pulse_samples(time=time)` or `pulse_samples(samples=samples)` and then error if both arguments (or none) are given.

I also exposed `dt` from the backend so you can do `timing.dt * timing.round_pulse_samples(samples)`. You could just use the backend object, but the class needed to carry dt as a member any way, so I thought it might as well allow access to it. I didn't want to try to think of a name for `round_pulse_samples_time`, though if the two keyword argument `pulse_samples` seems better, we could do a two keyword argument `pulse_time` as well.